config/feature_flags.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package config
+
+import (
+	"math"
+	"reflect"
+
+	"github.com/mattermost/mattermost-server/v5/mlog"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/pkg/errors"
+	"github.com/splitio/go-client/v6/splitio/client"
+	"github.com/splitio/go-client/v6/splitio/conf"
+)
+
+type FeatureFlagSyncParams struct {
+	ServerID            string
+	SplitKey            string
+	SyncIntervalSeconds int
+	Log                 *mlog.Logger
+}
+
+type FeatureFlagSynchronizer struct {
+	FeatureFlagSyncParams
+
+	client  *client.SplitClient
+	stop    chan struct{}
+	stopped chan struct{}
+}
+
+func NewFeatureFlagSynchronizer(params FeatureFlagSyncParams) (*FeatureFlagSynchronizer, error) {
+	cfg := conf.Default()
+	if params.Log != nil {
+		cfg.Logger = &splitMlogAdapter{wrappedLog: params.Log.With(mlog.String("service", "split"))}
+	} else {
+		cfg.LoggerConfig.LogLevel = math.MinInt32
+	}
+	factory, err := client.NewSplitFactory(params.SplitKey, cfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create split factory")
+	}
+
+	return &FeatureFlagSynchronizer{
+		FeatureFlagSyncParams: params,
+		client:                factory.Client(),
+		stop:                  make(chan struct{}),
+		stopped:               make(chan struct{}),
+	}, nil
+}
+
+// ensureReady blocks until the syncronizer is ready to update feature flag values
+func (f *FeatureFlagSynchronizer) EnsureReady() error {
+	if err := f.client.BlockUntilReady(10); err != nil {
+		return errors.Wrap(err, "split.io client could not initalize")
+	}
+
+	return nil
+}
+
+func (f *FeatureFlagSynchronizer) UpdateFeatureFlagValues(base model.FeatureFlags) model.FeatureFlags {
+	featureNames := getStructFields(model.FeatureFlags{})
+	featuresMap := f.client.Treatments(f.ServerID, featureNames, nil)
+	ffm := featureFlagsFromMap(featuresMap, base)
+	return ffm
+}
+
+func (f *FeatureFlagSynchronizer) Close() {
+	f.client.Destroy()
+}
+
+// featureFlagsFromMap sets the feature flags from a map[string]string.
+// It starts with baseFeatureFlags and only sets values that are
+// given by the upstream managment system.
+// Makes the assumption that all feature flags are strings for now.
+func featureFlagsFromMap(featuresMap map[string]string, baseFeatureFlags model.FeatureFlags) model.FeatureFlags {
+	refStruct := reflect.ValueOf(&baseFeatureFlags).Elem()
+	for fieldName, fieldValue := range featuresMap {
+		refField := refStruct.FieldByName(fieldName)
+		// "control" is returned by split.io if the treatment is not found, in this case we should use the default value.
+		if !refField.IsValid() || !refField.CanSet() || fieldValue == "control" {
```

Comment:
So does this mean we cannot have a feature flag value as "control"? Should we document this somewhere?

Reply:
Correct, there are some reserved values in the split.io system. No need for docs because the management system won't let you set the value wrong.